ipaserver/plugins/baseuser.py

```diff
@@ -728,3 +739,138 @@ def post_callback(self, ldap, dn, entry_attrs, *keys, **options):
         self.obj.convert_usercertificate_post(entry_attrs, **options)
 
         return dn
+
+
+class ModCertMapData(LDAPModAttribute):
+    attribute = 'ipacertmapdata'
+    takes_options = (
+        DNParam(
+            'issuer?',
+            cli_name='issuer',
+            label=_('Issuer'),
+            doc=_('Issuer of the certificate'),
+            flags=['virtual_attribute', 'no_create', 'no_update']
+        ),
+        DNParam(
+            'subject?',
+            cli_name='subject',
+            label=_('Subject'),
+            doc=_('Subject of the certificate'),
+            flags=['virtual_attribute', 'no_create', 'no_update']
+        ),
+        Bytes(
+            'usercertificate*', validate_certificate,
+            cli_name='certificate',
+            label=_('Certificate'),
+            doc=_('Base-64 encoded user certificate'),
+        ),
+    )
+
+    @staticmethod
+    def _build_mapdata(subject, issuer):
+        return u'X509:{issuer}{subject}'.format(
+            issuer='<I>{}'.format(issuer.x500_text()) if issuer else '',
+            subject='<S>{}'.format(subject.x500_text()) if subject else '',
+            )
+
+    @staticmethod
+    def _convert_options_to_certmap(certmapdata, options):
+        """
+        Converts options to ipacertmapdata
+
+        When --subject --issuer or --usercertificate options are used,
+        the value for ipacertmapdata is built from extracting subject and
+        issuer,
+        converting their values to X500 ordering and using the format
+        X509:<I>issuer<S>subject
+        For instance:
+        X509:<I>O=DOMAIN,CN=Certificate Authority<S>O=DOMAIN,CN=user
+        A list of values can be returned if --usercertificate is used multiple
+        times, or in conjunction with --subject --issuer.
+        """
+        data = []
+        data.extend(certmapdata)
+
+        if 'issuer' in options or 'subject' in options:
+            issuer = options.get('issuer')
+            subject = options.get('subject')
+            data.append(ModCertMapData._build_mapdata(subject, issuer))
```

Comment:
IMO it would be better to declare `_convert_options_to_certmap()` as class method, so you can `cls._build_mapdata()` instead of `ModCertMapData._build_mapdata()`.